### PR TITLE
data() tests & fixes for #48, dim() test tweaks

### PR DESCRIPTION
**Pull request by rvagg**
ref #48

### CODE REVIEW

**Comment by ded** [comment]
fyi: nice tip for writing commits. type `closes #48` in your commit... so that when they get merged, it will automatically close out the issues

**Comment by rvagg** [comment]
oh, that's neat, will try and remember.

**Comment by rvagg** [comment]
bump

**Comment by rvagg** [comment]
I'm going to keep pushing to this branch until you merge or close this PR!

Diff looks scarier than it is. You'll notice that the major changes in _src/_ are still to fix my original buggy `.data()` implementation. The other changes are to fix all the cases where Bonzo doesn't do the right thing when one of its functions are called on an empty collection (see #42). There is a big new test set for empty-collection cases that tests absolutely everything and will even test new functions that are introduced after this commit. Functions that take one or more arguments require test-data to be entered so it know what dummy values to pass in and how to check the result. No-arg functions don't need data if they just return `this` (like `.show()`), if they return something else then they need test data.

I've also now included an Ender integration into the test suite, the integration tests don't cover enough given how much new stuff is introduced in the bridge. For now I'm just running an Ender object with an empty collection through the same tests (so it tests `bonzo([]).*()` and also `ender([]).*()`). There are a few functions on Ender that don't exist on Bonzo and they are picked up properly.

You'll see a bunch of fixes in both _bonzo.js_ and _ender.js_ that have arisen from these new tests, the cases where errors are thrown or the wrong thing happens.

Merging this is going to close 3 issues, is that tempting enough?

**Comment by rvagg** [comment]
Also, this is related: https://github.com/ender-js/ender-js/pull/8/files

**Comment by ded** [comment]
ok, this one is not merging now.... and i was finally gonna have a hard look at it

**Comment by rvagg** [comment]
@nzakas mentioned a neat _NoScope_ tag `innerHTML` insertion hack on a recent [blog post](http://www.nczonline.net/blog/2012/01/03/css-media-queries-in-javascript-part-1/) which is perfect for getting more elements through `bonzo.create()` so I've included those. The only remaining HTML4 tags we can't insert are FRAME FRAMESET HEAD HTML ISINDEX TITLE, plus IE6 doesn't like ABBR and FF 3.6 and below don't like AREA.

Plus I fixed an annoying test-failure bug on IE8 and below that kept on showing up when you reload the test page. Narrowed it down to a stupid domready timing form cache issue so the form `reset()` is right with the test now.

Plus the recent `append()` hack for inserting the likes of a `<p/>` into a `<p/>` needed a `toLowerCase()` to actually pass in IE even though the fix was for IE and is working fine!

Plus some other minor items.

**Comment by rvagg** [comment]
Fixed another bug in `data()`. When attributes resolves to a falsey value it would come back as `null` when requested by name (`.data('somekey')`, but fine for `.data().somekey`).
I've also changed the API rule that non-existant data keys return `null`, they now return `undefined` which makes more sense (and takes up less code to implement).

**Comment by rvagg** [comment]
Latest commit completes the existing partial support for space separated lists of class names in add/remove/hasClass.

**Comment by rvagg** [comment]
Fixed `height(0)` and `width(0)` in ender.js, it was just a falsey check. Now able to test the bridge in the tests btw, `$` is raw Bonzo and `ender` is Bonzo fitted to Ender.

**Comment by rvagg** [comment]
changed `.height()` and `.width()` in ender.js so they now use `dim()` instead of their own custom function.

**Comment by ded** [comment]
damn damn. so this is what happens when putting off a PR this long. this will be some good diff reading

File: src/bonzo.js [hunk 1]
```diff
@@ -101,9 +105,13 @@
     uid && (delete uidMap[uid])
   }
 
-  function dataValue(d) {
+  function dataValue(d, f) {
     try {
-      return d === 'true' ? true : d === 'false' ? false : d === 'null' ? null : !isNaN(d) ? parseFloat(d) : d;
+      return (d === null || d === undefined) ? undefined :
```
Comment (ded) [inline src/bonzo.js:110]: jesus i wish there was a better way to write this — but this works

File: src/bonzo.js [hunk 2]
```diff
@@ -112,7 +120,7 @@
     return node && node.nodeName && node.nodeType == 1
   }
 
-  function some(ar, fn, scope, i) {
+  function some(ar, fn, scope, i, j) {
```
Comment (ded) [inline src/bonzo.js:123]: i love that my style is catching on by passing in vars that get redefined :)

File: src/bonzo.js [hunk 3]
```diff
@@ -210,21 +218,25 @@
   // so we have to iterate. bullshit
   if (features.classList) {
     hasClass = function (el, c) {
-      return some(c.toString().split(' '), function (c) {
```
Comment (ded) [inline src/bonzo.js:213]: why did you revert this? please undo this change — otherwise `$('p').hasClass('foo bar')` will not work

File: src/bonzo.js [hunk 4]
```diff
@@ -210,21 +218,25 @@
...
-        return el.classList.contains(c)
-      })
+      return el.classList.contains(c)
     }
     addClass = function (el, c) {
-      each(c.toString().split(' '), function (c) {
```
Comment (ded) [inline src/bonzo.js:218]: same with this one. `$('p').addClass('foo bar')`

File: src/bonzo.js [hunk 5]
```diff
@@ -379,28 +391,44 @@
 
       // class management
     , addClass: function (c) {
+        c = toString.call(c).split(whitespaceRegex)
```
Comment (ded) [inline src/bonzo.js:394]: ok. i see. moved the logic here.... right on

File: src/bonzo.js [hunk 6]
```diff
@@ -443,8 +471,8 @@
       }
 
     , parent: function() {
-      return this.related('parentNode')
-    }
+        return this.related(parentNode)
```
Comment (ded) [inline src/bonzo.js:474]: sweet

File: src/bonzo.js [hunk 7]
```diff
@@ -463,7 +491,8 @@
 
       // meh. use with care. the ones in Bean are better
     , focus: function () {
-        return this.length > 0 ? this[0].focus() : null
```
Comment (ded) [inline src/bonzo.js:466]: yikes. this is what it was before? good change.

File: tests/tests.html [hunk 8]
```diff
@@ -99,9 +107,6 @@
           <input type="text" value="eyo">
           <input type="checkbox" checked>
         </form>
-        <script type="text/javascript">
-          document.forms[0].reset() // IE sometimes caches the value we set on the input between reloads
```
Comment (ded) [inline tests/tests.html:103]: what was the deal with this? **edit** ok, i see the logic moved

File: tests/tests.html [hunk 9]
```diff
@@ -157,14 +162,72 @@
           ok(dom(el).hasClass('existing-class'), 'element still has pre-existing class "existing-class"');
         });
 
+        test('add/remove/has multiple classes', 12 + 6 * 4, function () {
```
Comment (ded) [inline tests/tests.html:165]: 12 + 6 \* 4 = lol

File: tests/tests.html [hunk 10]
```diff
@@ -345,20 +408,21 @@
...
           var $el = $(dom.create('<div/>')).css({
-                position: 'absolute',
-                left: '50px',
-                top: -999,
-                height: '30px',
-                width: '40px',
-                border: 'solid 2px red'
+                  position: 'absolute'
+                , left: '50px'
+                , top: '-999px'
```
Comment (ded) [inline tests/tests.html:413]: horray comma first

File: tests/tests.html [hunk 11]
```diff
@@ -684,8 +737,238 @@
...
+              }
+            , isUndefined = function (o) { return o === undefined }
+            , isNull = function (o) { return o === null }
+            , isFalse = function (o) { return o === false }
+              // an empty function
+            , K = function () {}
+
+
+          //-----------------------------------------------------------------------------------------
+          // This is the most important bit. We need a definition here for each Bonzo/Ender function
```
Comment (ded) [inline tests/tests.html:802]: wow. now these are some muthafukin tests. damn
